control/discovery.py

```diff
@@ -0,0 +1,1077 @@
+            return -1
+        self.logger.debug("reply connect request.")
+        return 0
+
+    def reply_fc_cmd_prop_get(self, conn, data, cmd_id):
+        """Reply property get request."""
+
+        self.logger.debug("handle property get request.")
+        self_conn = self.conn_vals[conn.fileno()]
+        nvmeof_prop_get_set_rsvd0 = struct.unpack_from('<35B', data, 13)
+        # property size = (attrib+1)x4, 0x1 means 8 bytes
+        nvmeof_prop_get_set_attrib = struct.unpack_from('<1B', data, 48)[0]
+        nvmeof_prop_get_set_rsvd1 = struct.unpack_from('<3B', data, 49)
+        nvmeof_prop_get_set_offset = struct.unpack_from('<I', data, 52)[0]
+
+        pdu_reply = Pdu()
+        pdu_reply.type = NVME_TCP_PDU.RSP
+        pdu_reply.header_length = 24
+        pdu_reply.packet_length = 24
+
+        # reply different property data
+        property_get = CqePropertyGetSet()
+        if NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.CAPABILITIES:
+            # controller capabilities
+            # \x7f = maxinum queue entries support:128
+            # \x01 contiguous queues required: true
+            # \x1e timeout(to ready status): 1e(15000ms default in server side)
+            # \x20 command sets supportd: 1 (NVM IO command set)?
+            property_get.property_data = (c_ubyte * 8)(0x7f, 0x00, \
+                0x01, 0x1e, 0x20, 0x00, 0x00, 0x00)
+        elif NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.CONFIGURATION:
+            # b'\x00\x00\x46\x00\x00\x00\x00\x00'
+            # 0x46: IO Submission Queue Entry Size: 0x6 (64 bytes)
+            # IO Completion Queue Entry Size: 0x4 (16 bytes)
+            property_data = (c_ubyte * 8)(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+            if len(self_conn.property_data) == 8:
+                property_data = (c_ubyte * 8)(*self_conn.property_data)
+            property_get.property_data = property_data
+        elif NVME_CTL(nvmeof_prop_get_set_offset) == NVME_CTL.STATUS:
+            property_data = (c_ubyte * 8)(0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+            if len(self_conn.property_data) == 8:
+                property_data = (c_ubyte * 8)(*self_conn.property_data)
+            shutdown_notification = ((property_data[1]) >> 6) & 0x3
+            if shutdown_notification == 0:
+                if self_conn.property_set is True:
+                    # controller status: ready
+                    property_get.property_data = (c_ubyte * 8)(0x01, 0x00, \
+                        0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+                else:
+                    property_get.property_data = (c_ubyte * 8)(0x00, 0x00, \
+                        0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+            else:
+                # here shutdown_notification should be 0x1
+                property_get.property_data = (c_ubyte * 8)(0x09, 0x00, \
+                    0x00, 0x00, 0x00, 0x00, 0x00, 0x00)
+                self_conn.shutdown_now = True
```

Comment:
Should we close the connection here?

According to spec, the controller should not initiate a disconnect
at the NVMe Transport level:

```
After the host specifies a shutdown, the
host may either disconnect at the NVMe Transport level or it
may choose to poll CSTS.SHST to determine when the shutdown
is complete (the controller should not initiate a disconnect
at the NVMe Transport level).
```

Reply:
Here we should close connection. The meaning of spec is that the discovery controller will not send a disconnect command, and in fact, it will not. It does not have a disconnect command in the protocol. This line is where the shutdown command is received, then saved, waiting for the next `STATUS` message to confirm, and finally closing the connection, which is passive.